File: 0621-task-scheduler/0621-task-scheduler.js
var leastInterval = function (tasks, n) {
    // Frequency of each character
    let freq = new Array(26).fill(0);
    for (let t of tasks) freq[t.charCodeAt(0) - 65]++;
    let heap = freq.filter(x => x > 0);

    let time = 0;
    while (heap.length > 0) {
        // To collect the tasks for this cycle
        let temp = [];
        let tasksDone = 0;
        for (let i = 0; i <= n; i++) {
            if (heap.length > 0) {
                temp.push(heap.shift());
                tasksDone++;
            } else {
                temp.push(0); // pad with zero for completeness
            }
        }
        // Put back those with remaining count
        for (let count of temp) {
            if (count > 1) heap.push(count - 1);
        }
        // Re-sort to maintain max-heap behaviour
        heap.sort((a, b) => b - a);

        // If heap is empty after this round, don't add idle time
        time += heap.length === 0 ? tasksDone : (n + 1);
    }
    return time;
};

File: 0621-task-scheduler/0621-task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        # technically take the empty space as needed and fill them up
        # so maxCount - 1 * (n+1) rows + last row with characters of max Count (since each char should have n gap)
        counts = Counter(tasks).values()
        maxCount = max(counts)
        charCountWithMaxCount = sum(1 for c in counts if c == maxCount)
        
        #actually there can be many takss which may overflow our grid especially when maxCount is small and n is small, so we need to check the task lengthn as well
        # When there are so many tasks that they naturally fill all idle slots and overflow, no idle time is needed, you just run all tasks back to back.
        return max(len(tasks), (maxCount - 1) * (n + 1) + charCountWithMaxCount)